File: tooling/harness-eval/codex.ts
import type { Run } from "./report-schema.ts";
import { z } from "zod";

const eventSchema = z.looseObject({ type: z.string() }).readonly();
const usageSchema = z.object({
  usage: z.object({
    input_tokens: z.int().nonnegative(),
    cached_input_tokens: z.int().nonnegative(),
    output_tokens: z.int().nonnegative(),
  }),
});
const itemSchema = z.object({
  item: z.object({ id: z.string(), type: z.string() }).readonly(),
});

type CodexMetrics = Readonly<{
  tokens: NonNullable<Run["tokens"]>;
  toolCalls: number;
}>;

function parseCodexEvents(output: string): CodexMetrics {
  const events = output
    .split("\n")
    .filter(Boolean)
    .map((line) => eventSchema.parse(JSON.parse(line)));
  if (
    events.some(
      (event) => event.type === "error" || event.type === "turn.failed",
    )
  ) {
    throw new Error("Agent failure event");
  }
  const completed = events.filter((event) => event.type === "turn.completed");
  if (completed.length !== 1) {
    throw new Error("One completed turn required");
  }
  const { usage } = usageSchema.parse(completed[0]);
  const calls = events
    .filter((event) => event.type === "item.completed")
    .map((event) => itemSchema.parse(event).item)
    .filter((item) =>
      [
        "command_execution",
        "mcp_tool_call",
        "web_search",
        "file_change",
      ].includes(item.type),
    );
  return {
    tokens: {
      input: usage.input_tokens,
      cachedInput: usage.cached_input_tokens,
      output: usage.output_tokens,
    },
    toolCalls: new Set(calls.map((item) => item.id)).size,
  };
}

export { type CodexMetrics, parseCodexEvents };

File: tooling/arnes/src/eval/codex.rs
use serde::Deserialize;
use std::collections::BTreeSet;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Tokens {
    pub input: u64,
    pub cached_input: u64,
    pub output: u64,
}

#[derive(Debug)]
pub struct CodexMetrics {
    pub tokens: Tokens,
    pub tool_calls: u64,
}

#[derive(Deserialize)]
struct Event {
    #[serde(rename = "type")]
    kind: String,
    #[serde(flatten)]
    fields: serde_json::Map<String, serde_json::Value>,
}

#[derive(Deserialize)]
struct Usage {
    #[serde(rename = "input_tokens")]
    input: f64,
    #[serde(rename = "cached_input_tokens")]
    cached_input: f64,
    #[serde(rename = "output_tokens")]
    output: f64,
}

#[derive(Deserialize)]
struct Item {
    id: String,
    #[serde(rename = "type")]
    kind: String,
}

/// # Errors
/// Rejects malformed events, failed turns, missing or repeated completion events, and invalid token usage.
pub fn parse_codex_events(output: &str) -> Result<CodexMetrics, String> {
    let mut tokens = None;
    let mut calls = BTreeSet::new();
    for line in output.split('\n').filter(|line| !line.is_empty()) {
        let event: Event = serde_json::from_str(line).map_err(|error| error.to_string())?;
        match event.kind.as_str() {
            "error" | "turn.failed" => return Err("Agent failure event".into()),
            "turn.completed" => {
                if tokens.is_some() {
                    return Err("One completed turn required".into());
                }
                let usage: Usage = event_field(&event, "usage")?;
                let input = crate::numbers::safe_integer(usage.input);
                let cached_input = crate::numbers::safe_integer(usage.cached_input);
                let output = crate::numbers::safe_integer(usage.output);
                let (Some(input), Some(cached_input), Some(output)) = (input, cached_input, output)
                else {
                    return Err("Usage exceeds safe integer range".into());
                };
                let validated = Tokens {
                    input,
                    cached_input,
                    output,
                };
                tokens = Some(validated);
            }
            "item.completed" => {
                let item: Item = event_field(&event, "item")?;
                if [
                    "command_execution",
                    "mcp_tool_call",
                    "web_search",
                    "file_change",
                ]
                .contains(&item.kind.as_str())
                {
                    calls.insert(item.id);
                }
            }
            _ => {}
        }
    }
    Ok(CodexMetrics {
        tokens: tokens.ok_or("One completed turn required")?,
        tool_calls: calls.len() as u64,
    })
}

fn event_field<T: serde::de::DeserializeOwned>(event: &Event, name: &str) -> Result<T, String> {
    let value = event
        .fields
        .get(name)
        .ok_or_else(|| format!("Missing {name}"))?;
    serde_json::from_value(value.clone()).map_err(|error| error.to_string())
}

#[cfg(test)]
mod tests;
